storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageClient.java

```diff
@@ -306,4 +306,46 @@ public void onNewSlot(SlotEvent slotEvent) {
       }
     }
   }
+
+  public Optional<BeaconBlock> getBlockBySlot(final UnsignedLong slot) {
+    return getBlockRootBySlot(slot).map(blockRoot -> store.getBlock(blockRoot));
+  }
+
+  public boolean isIncludedInBestState(final Bytes32 blockRoot) {
+    if (store == null) {
+      return false;
+    }
+    final BeaconBlock block = store.getBlock(blockRoot);
+    if (block == null) {
+      return false;
+    }
+    return getBlockRootBySlot(block.getSlot())
+        .map(actualRoot -> actualRoot.equals(block.hash_tree_root()))
+        .orElse(false);
+  }
+
+  private Optional<Bytes32> getBlockRootBySlot(final UnsignedLong slot) {
+    if (store == null || Bytes32.ZERO.equals(bestBlockRoot)) {
+      return Optional.empty();
+    }
+    final UnsignedLong slotsPerHistoricalRoot =
+        UnsignedLong.valueOf(Constants.SLOTS_PER_HISTORICAL_ROOT);
+
+    if (isSlotStillAvailable(slot, slotsPerHistoricalRoot)) {
```

Comment:
Not sure if its worth it - but we could take the oldest available root from best block's state, and then keep paging back through history as long as its available.

Reply:
We'll need to fall back to getting the block from storage at some point anyway which should have the full weak subjectivity period available. Hopefully it can do it with an efficient index rather than having to walk back.